src/librustc/ty/context.rs

```diff
@@ -481,6 +481,8 @@ pub struct GlobalCtxt<'tcx> {
 
     pub maybe_unused_trait_imports: NodeSet,
 
+    pub maybe_unused_extern_crates: Vec<(NodeId, Span, CrateNum)>,
```

Comment:
You can use `NodeSet` here, like `maybe_unused_trait_imports` does.
Crate num can be obtained from node id (`tcx.sess.cstore.extern_mod_stmt_cnum(item.id)`).
In `check_unused.rs` extern crate checking can be done in the same visitor where `check_import` lives (it will also fix the issue with reporting `pub` extern crates as unused, which seems to exist in the current implementation).